webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -1125,211 +1298,138 @@ final case class ContextForNoAnswer(
-                    val card = value.cardinality + (if (incrementCardinality(value, report)) 1 else 0)
-                    val (r, nbDup) = if (dropDuplicated(value, report)) {
-                      val msg = s"Following report is duplicated and will be ignored because of Rudder setting choice: ${report.toString}"
-                      if (value.numberDuplicates <= 0) { //first time is an info
-                        logger.info(msg)
-                        (value.matchingReports, 1)
-                      } else if (value.numberDuplicates == 1) { //second time is a warning
-                        logger.warn(msg)
-                        (value.matchingReports, 2)
-                      } else { // more than two times: log error and let the report leads to an unexpected
-                        val n = value.numberDuplicates + 1
-                        logger.error(s"Following report is duplicated ${n} times. This is spurious and should be investigated. The message is kept as unexpected despite Rudder setting.")
-                        (report :: value.matchingReports, n)
-                      }
-                    } else {
-                      (report :: value.matchingReports, value.numberDuplicates)
-                    }
-                    (stack, Some(value.copy(
-                      cardinality = card
-                      , numberDuplicates = nbDup
-                      , matchingReports = r
-                    )))
-                  } else {
-                    (value :: stack, None)
-                  }
-              }
-            }
-            (stack.reverse, found)
-          }
+    def checkExpectedBlock(
+        expectedComponent       : BlockExpectedReport
+      , directiveFilteredReports: Seq[ResultReports]
+      , noAnswerType            : ReportType
+      , policyMode              : PolicyMode //the one of the directive, or node, or global
+      , unexpectedInterpretation: UnexpectedReportInterpretation
+    ): BlockStatusReport = {
+      def recFlattenExpectedComponent(component: ComponentExpectedReport): List[ValueExpectedReport] = {
+         component match {
+          case g: BlockExpectedReport => g.subComponents.flatMap(recFlattenExpectedComponent)
+          case v: ValueExpectedReport => v :: Nil
+        }
+      }
 
-          reports match {
-            case Nil => (freeValues ::: pairedValues, unexpected)
-            case report :: tail =>
-              // we look in the free values for one matching the report. If found, we return (remaining freevalues, Some[paired value]
-              // else (all free values, None).
-              // never increment the cardinality for free value.
-              val (newFreeValues, tryPair) = findMatchingValue(report, freeValues, (value, report) => false, (value, report) => false)
+      def buildBackTree(expectedComponent: BlockExpectedReport, valueReports: List[ValueStatusReport]): BlockStatusReport = {
```

Comment:
This is the complicated part that remains. We need to unmerge values to build back the correct tree of component blocks.